### PR TITLE
Change mailing list subscription to when the user validates their email

### PR DESCRIPTION
**Pull request by ericholscher**
This change also adds support for subscribing the user to an onboarding group.
Users in this group will get custom onboarding message(s),
to help them along on their usage of the platform.

Refs https://github.com/readthedocs/readthedocs-corporate/issues/1394

### CODE REVIEW

File: readthedocs/core/signals.py [hunk 1]
```diff
@@ -33,6 +34,54 @@
...
+    user = email_address.user
+    profile = UserProfile.objects.filter(user=user).first()
+    if profile and profile.mailing_list:
+        log.bind(
+            email=email_address.email,
+            username=user.username,
+        )
+        log.info("Subscribing user to newsletter and onboarding group.")
+
+        # Try subscribing used to a group, then fallback to normal subscription API
```
Comment (stsewd) [inline readthedocs/core/signals.py:61]:
```suggestion
        # Try subscribing user to a group, then fallback to normal subscription API
```

File: readthedocs/core/models.py [hunk 2]
```diff
@@ -20,33 +16,34 @@ class UserProfile(TimeStampedModel):
...
     """Additional information about a User."""
 
-    # TODO: Overridden from TimeStampedModel just to allow null values,
-    # remove after deploy.
-    created = CreationDateTimeField(
-        _('created'),
-        null=True,
-        blank=True,
-    )
-    modified = ModificationDateTimeField(
```
Comment (stsewd) [inline readthedocs/core/models.py:30]: we need to create a data migration to migrate any null values, we could also just do this from the admin before running the migration

File: readthedocs/settings/base.py [hunk 3]
```diff
@@ -906,6 +906,7 @@ def DOCKER_LIMITS(self):
 
     # MailerLite API for newsletter signups
     MAILERLITE_API_SUBSCRIBERS_URL = 'https://api.mailerlite.com/api/v2/subscribers'
+    MAILERLITE_API_ONBOARDING_GROUP_URL = None
```
Comment (stsewd) [inline readthedocs/settings/base.py:909]: Are we going to use the other value somewhere else? Looks like we are using just one of these settings.

Comment (ericholscher) [reply readthedocs/settings/base.py:909]: Yea, it's just group specific so didn't get included in the PR.

File: readthedocs/core/signals.py [hunk 4]
```diff
@@ -33,6 +36,54 @@
...
+            "X-MailerLite-ApiKey": settings.MAILERLITE_API_KEY,
+        }
+        try:
+            resp = requests.post(
+                url,
+                json=payload,
+                headers=headers,
+                timeout=3,  # seconds
+            )
+            resp.raise_for_status()
```
Comment (humitos) [inline readthedocs/core/signals.py:80]: Shouldn't this be done in a Celery task?

**Review by humitos — APPROVED** [review]
This is a good improvement to reduce the spam emails added to our newsletter! 💯

File: readthedocs/core/signals.py (hunk 4 above)
Comment (ericholscher) [reply readthedocs/core/signals.py:80]: Probably. I was just copying the existing logic for now 🙃. There's definitely more work to be done here, including handling the updating of this on save to unsub the user.

**Comment by humitos** [comment]
I found this PR today and looks useful. I'm commenting here to bump it up in your notifications, @ericholscher. It would be good to update if needed and merge it.

**Comment by ericholscher** [comment]
@humitos Yea, I've had it open in a tab for a couple weeks now trying to get back to it. It's still on my roadmap, just need to find the time to finish it up.

**Comment by humitos** [comment]
@ericholscher 

I re-read this PR today and I'm not sure how it works on the MailerLite side and what use cases exactly it covers.

I understand that we have a "Onboard" (or similar) group on MailerLite that we add new users to it when the verify their email. At this point, MailerLite, _somehow_ sends them the first onboarding email:

- how does it differentiate from new users and old users of that group?
- how do we tell MailerLite to send emails to these new users, but after 3 days they have subscribed?

I may be confusing some things here and it may be easy for you to explain me this in a call. As you prefer.

**Comment by ericholscher** [comment]
@humitos The tooling has support for sending campaigns to users who have signed up in a specific sequence. I don't really understand your questions though? We just only send the campaign when users are added, and it has a specific schedule.

**Comment by humitos** [comment]
All doubts are cleared out after chatting with you, 👍🏼

File: readthedocs/core/models.py (hunk 2 above)
Comment (humitos) [reply readthedocs/core/models.py:30]: I created a data migration for this.

File: readthedocs/settings/base.py (hunk 3 above)
Comment (humitos) [reply readthedocs/settings/base.py:909]: @ericholscher Do we need to create an -ops PR for this?

File: readthedocs/core/signals.py (hunk 4 above)
Comment (humitos) [reply readthedocs/core/signals.py:80]: I added a TODO note for now.

**Comment by humitos** [comment]
@ericholscher I updated this PR with the migration required. Let me know if there is something missing here. I'd love to deploy this next week so we can start sending emails to new users.

File: readthedocs/settings/base.py (hunk 3 above)
Comment (ericholscher) [reply readthedocs/settings/base.py:977]: https://github.com/readthedocs/readthedocs-ops/pull/1254

**Review by ericholscher — COMMENTED** [review]
These changes look good. Excited to ship it 👍

File: readthedocs/core/signals.py [hunk 5]
```diff
@@ -33,6 +36,55 @@
...
+
+    * Subscribe the user to the mailing list if they set this in their signup.
+    * Add them to a drip campaign for new users in the new user group.
+    """
+
+    # email_address is an object
+    # https://github.com/pennersr/django-allauth/blob/6315e25/allauth/account/models.py#L15
+    user = email_address.user
+    profile = UserProfile.objects.filter(user=user).first()
+    if profile and profile.mailing_list:
```
Comment (ericholscher) [inline readthedocs/core/signals.py:54]: I think we aren't unsubscribing people in this PR still, which we should be doing, but I think we can do it later. 

```suggestion
    if profile and profile.mailing_list:
    # TODO: Unsubscribe users if they unset `mailing_list`.
```

**Pull request merged**